File: rust-2020/src/day09.rs
use itertools::Itertools;
use itertools::MinMaxResult::MinMax;

pub fn read(contents: &str) -> Vec<i64> {
    contents
        .lines()
        .map(|line| line.parse::<i64>().unwrap())
        .collect::<Vec<_>>()
}

pub fn part1(data: &[i64]) -> i64 {
    let window_size = 25;
    *data[window_size..]
        .iter()
        .zip(data.windows(window_size))
        .find(|(&want, have)| !(*have).iter().any(|h| have.contains(&(want - h))))
        .unwrap()
        .0
}

pub fn part2(data: &[i64]) -> i64 {
    let want = part1(data);
    let mut left = 0;
    let mut subseq_sum = 0;
    for (right, n) in data.iter().enumerate() {
        while subseq_sum > want && left < right {
            subseq_sum -= data[left];
            left += 1;
        }
        if subseq_sum == want {
            if let MinMax(&min, &max) = &data[left..right].iter().minmax() {
                return min + max;
            }
        }
        subseq_sum += n;
    }
    unreachable!()
}
pub fn run() -> (String, String) {
    let contents = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/inputs/day09.input"));
    let data = read(contents);
    let p1 = part1(&data).to_string();
    let p2 = part2(&data).to_string();
    (p1, p2)
}

File: 2020/day09.py
from more_itertools import windowed

with open("../inputs/2020/day09.input", "r") as f:
    data = list(map(int, f.readlines()))


def part1(data, window_size=25):
    return next(want for want, have in
                zip(data[window_size:], map(set, windowed(data, window_size)))
                if not any(want - h in have for h in have))


def part2(data, want):
    data = data[:data.index(want)]
    left = 0
    right = 0
    subseq_sum = 0
    for right, n in enumerate(data):
        while subseq_sum > want and left < right:
            subseq_sum -= data[left]
            left += 1
        if subseq_sum == want:
            subseq = data[left:right]
            return min(subseq) + max(subseq)
        subseq_sum += n
    raise RuntimeError("")


want = part1(data)
print(f"Part 1: {want}")
print(f"Part 2: {part2(data, want)}")
